### PR TITLE
Fix check for private key existence before eddsa sign functions

### LINKED ISSUE #19524: ED25519_sign and ED448_sign missing check for private_key leading to segfault
Consider the following example:

```c
#include <assert.h>
#include <stdio.h>
#include <string.h>
#include <openssl/crypto.h>
#include <openssl/evp.h>
#include <openssl/x509.h>


int main(void) {
    OPENSSL_init();
    int ret = 0;
    int res;
    unsigned char *public = OPENSSL_hexstr2buf("7d4d0e7f6153a69b6242b522abbee685fda4420f8834b108c3bdae369ef549fa", NULL);
    unsigned char *msg = OPENSSL_hexstr2buf("54657374", NULL);


    size_t sig_len;
    unsigned char *sigg = NULL;

    EVP_PKEY *pkey = EVP_PKEY_new_raw_public_key(NID_ED25519, NULL, public, 32);
    assert(pkey != NULL);
    EVP_MD_CTX *ctx = EVP_MD_CTX_new();
    res = EVP_DigestSignInit(ctx, NULL, NULL, NULL, pkey);
    assert(res == 1);

    if (EVP_DigestSign(ctx, NULL, &sig_len, msg, 4) != 1)
       printf("EVP_DigestSign get the size of signature buffer fail\n");
    sigg = OPENSSL_zalloc(sig_len);
    if (EVP_DigestSign(ctx, sigg, &sig_len, msg, 4) != 1)
       printf("EVP_DigestSign fail\n");

    printf("ok\n");
    exit(ret);
}
```

If someone accidently use a public key to sign, it will cause segfault. On openssl3.0.2 , it produces following stack trace:

```
#0  SHA512_Update (c=0x5555555a8090, _data=<optimized out>, len=32) at ../crypto/sha/sha512.c:322
#1  0x00007ffff7e0b091 in ossl_ed25519_sign.constprop.0 (out_sig=0x55555556b1a0 "", message=0x5555555592d0 "Test", message_len=4,
    public_key=0x555555598bd1 "}M\016\177aS\246\233bB\265\"\253\276\346\205\375\244B\017\210\064\261\bý\256\066\236\365", <incomplete sequence \372>,
    private_key=0x0, libctx=<optimized out>, propq=0x0, propq=0x0) at ../crypto/ec/curve25519.c:5460
#2  0x00007ffff7def1df in ed25519_digest_sign (vpeddsactx=<optimized out>, sigret=<optimized out>, siglen=0x7fffffffe518, sigsize=<optimized out>,
    tbs=<optimized out>, tbslen=<optimized out>) at ../providers/implementations/signature/eddsa_sig.c:171
#3  0x00005555555553e2 in main () at aa.c:29
```

The problem also exists for Ed448_sign; and in openssl 1.1.1r.

### PR DESCRIPTION
**Pull request by fffonion**
Fix #19524.

Looks like there's no test for cases of RSA and EC, please advice if I need to add test for this patch. Thanks.

A similar fix is also needed for 1.1.1, please let me know if this one looks good so I can backport.

##### Checklist

- [x] tests are added or updated

### CODE REVIEW

**Review by hlandau — DISMISSED** [review]

**Comment by slontis** [comment]
It should have a test..

**Comment by slontis** [comment]
A related question is where EVP_DigestUpdate() should segfault if it is passed NULL as its data..

**Comment by mattcaswell** [comment]
> It should have a test..

Yes a test should be there for this. Perhaps in test/evp_extra_test.c

**Comment by fffonion** [comment]
I will add test tommorow, thanks for reviewing!

File: providers/implementations/signature/eddsa_sig.c [hunk 1]
```diff
@@ -161,6 +162,10 @@ int ed25519_digest_sign(void *vpeddsactx, unsigned char *sigret,
         ERR_raise(ERR_LIB_PROV, PROV_R_OUTPUT_BUFFER_TOO_SMALL);
         return 0;
     }
+    if (edkey->privkey == NULL) {
+        ERR_raise(ERR_LIB_PROV, EC_R_MISSING_PRIVATE_KEY);
```
Comment (t8m) [inline providers/implementations/signature/eddsa_sig.c:166]: This is wrong, you cannot use EC_R with ERR_LIB_PROV. Use `PROV_R_NOT_A_PRIVATE_KEY` instead.

File: providers/implementations/signature/eddsa_sig.c [hunk 2]
```diff
@@ -198,6 +203,10 @@ int ed448_digest_sign(void *vpeddsactx, unsigned char *sigret,
...
```
Comment (t8m) [inline providers/implementations/signature/eddsa_sig.c:207]: same here

**Review by t8m — CHANGES_REQUESTED** [review]

**Review by t8m — COMMENTED** [review]
This is also outside of limits for what we would accept with CLA: trivial. Could you please submit a regular CLA? https://www.openssl.org/policies/cla.html

**Comment by fffonion** [comment]
I have addressed the comments and sent CLA to legal. When I was running test locally, the pkey creation always fails with
```
routines:inner_evp_generic_fetch:unsupported:crypto/evp/evp_fetch.c:341:Global default library context, Algorithm (ED25519 : 114), Properties (<null>)
```
It seems the previous test `test_ecx_short_keys` is also failed with same error (although the key passed is not legit as well); I'm wondering if my local setup is wrong or if the test I wrote need to explictly load those algorithms?

File: test/evp_extra_test.c [hunk 3]
```diff
@@ -4620,14 +4620,24 @@ static int test_ecx_short_keys(int tst)
 {
     unsigned char ecxkeydata = 1;
     EVP_PKEY *pkey;
+    OSSL_PROVIDER* deflprov;
+    int testresult = 1;
 
 
+    deflprov = OSSL_PROVIDER_load(NULL, "default");
```
Comment (t8m) [inline test/evp_extra_test.c:4627]: This should not be necessary. Unless you have a configuration file that explicitly loads only the base and/or legacy provider which would be fairly broken configuration anyway.

None of the other tests explicitly load the default provider.

File: test/evp_extra_test.c [hunk 4]
```diff
@@ -4647,6 +4657,80 @@ const OPTIONS *test_get_options(void)
     return options;
 }
 
+/* Test that trying to sign with a public key errors out gracefully */
+static int test_ecx_not_private_key(int tst)
+{
+#ifdef OPENSSL_NO_EC
+    return 1;
+#else
```
Comment (t8m) [inline test/evp_extra_test.c:4665]: I'd prefer if the whole function was in `#ifndef OPENSSL_NO_EC` and the ADD_ALL_TESTS was also ifndef-ed.

File: test/evp_extra_test.c [hunk 5]
```diff
@@ -4647,6 +4657,80 @@ const OPTIONS *test_get_options(void)
...
+    unsigned char *pubkey;
+    size_t pubkeylen;
+    OSSL_PROVIDER* deflprov;
+
+
+    switch (keys[tst].type) {
+        case NID_X25519:
+        case NID_X448:
+            /* Test does not apply, skip */
+            return 1;
```
Comment (t8m) [inline test/evp_extra_test.c:4683]: Use `return TEST_skip(.....);`

File: test/evp_extra_test.c [hunk 6]
```diff
@@ -4647,6 +4657,80 @@ const OPTIONS *test_get_options(void)
...
+    }
+
+    /* Check if this algorithm supports public keys */
+    if (keys[tst].pub == NULL)
+        return 1;
+
+    deflprov = OSSL_PROVIDER_load(NULL, "default");
```
Comment (t8m) [inline test/evp_extra_test.c:4690]: This should be again unnecessary.

File: test/evp_extra_test.c [hunk 7]
```diff
@@ -4647,6 +4657,80 @@ const OPTIONS *test_get_options(void)
...
+        goto err;
+
+    ret = EVP_DigestSignInit(ctx, NULL, NULL, NULL,
+                             pkey);
+
+    if (!TEST_true(ret))
+        goto err;
+
+    if (!TEST_true(EVP_DigestSign(ctx, NULL, &maclen, msg, sizeof(msg))))
+        goto err;
```
Comment (t8m) [inline test/evp_extra_test.c:4713]: In theory, the provider could fail the operation already at the EVP_DigestSignInit() or the EVP_DigestSign() call with NULL signature buffer. I am not sure it is good idea to cement this behavior by the test. Failure of either of these three calls should be accepted as correct.

**Review by t8m — CHANGES_REQUESTED** [review]

File: test/evp_extra_test.c (hunk 3 above)
Comment (fffonion) [reply test/evp_extra_test.c:4627]: It seems the there's a test case to use a non-default ctx, where the `null` provider is loaded (https://github.com/openssl/openssl/blob/12b1ae38aa4621a85b00f560a3e828f9a5386d04/test/evp_extra_test.c#L4746), although I'm able to pass all tests after remove that line, but I assume some tests are actually "masked" and only looks to be passed.

File: test/evp_extra_test.c (hunk 7 above)
Comment (fffonion) [reply test/evp_extra_test.c:4713]: I agree, theoritically those two functions would also fail. But I'm bit concerned if some other failure in those two functions will mask the test, as we don't actually test the error message. How do you think?

File: test/evp_extra_test.c (hunk 4 above)
Comment (fffonion) [reply test/evp_extra_test.c]: fixed

File: test/evp_extra_test.c (hunk 5 above)
Comment (fffonion) [reply test/evp_extra_test.c]: fixed

File: test/evp_extra_test.c (hunk 3 above)
Comment (t8m) [reply test/evp_extra_test.c]: You need to skip the test if testctx is not NULL.

File: test/evp_extra_test.c (hunk 6 above)
Comment (t8m) [reply test/evp_extra_test.c]: Instead skip the test if testcx is not NULL.

File: test/evp_extra_test.c (hunk 7 above)
Comment (t8m) [reply test/evp_extra_test.c:4709]: You could call ERR_peek_error() and test that there is the appropriate error reason code on the stack.

File: test/evp_extra_test.c (hunk 3 above)
Comment (fffonion) [reply test/evp_extra_test.c]: That's a good idea, I will update the tests.

File: test/evp_extra_test.c [hunk 8]
```diff
@@ -4622,11 +4623,15 @@ static int test_ecx_short_keys(int tst)
     EVP_PKEY *pkey;
 
 
+    if (testctx != NULL)
+        return TEST_skip("ecx keys not supported in null provider");
```
Comment (t8m) [inline test/evp_extra_test.c:4627]: The message should be: `test does not work with non-default context`

Hmm, you could actually use EVP_PKEY_new_raw_private_key_ex() function instead and pass the testctx there and the TEST_skip would not be needed then. I am sorry for not realizing this earlier.

File: test/evp_extra_test.c [hunk 9]
```diff
@@ -4647,6 +4652,74 @@ const OPTIONS *test_get_options(void)
...
+    const unsigned char msg[] = { 0x00, 0x01, 0x02, 0x03 };
+    int testresult = 0;
+    EVP_MD_CTX *ctx = NULL;
+    unsigned char *mac = NULL;
+    size_t maclen = 0;
+    unsigned char *pubkey;
+    size_t pubkeylen;
+
+    if (testctx != NULL)
+        return TEST_skip("ecx keys not supported in null provider");
```
Comment (t8m) [inline test/evp_extra_test.c:4669]: Same as above.

File: test/evp_extra_test.c [hunk 10]
```diff
@@ -4647,6 +4652,74 @@ const OPTIONS *test_get_options(void)
...
+
+    switch (keys[tst].type) {
+        case NID_X25519:
+        case NID_X448:
+            return TEST_skip("signing not supported for X25519/X448");
+    }
+
+    /* Check if this algorithm supports public keys */
+    if (keys[tst].pub == NULL)
+        return 1;
```
Comment (t8m) [inline test/evp_extra_test.c:4679]: Again better use `TEST_skip("no public key present");`

File: test/evp_extra_test.c [hunk 11]
```diff
@@ -4647,6 +4652,74 @@ const OPTIONS *test_get_options(void)
...
+
+    pubkey = (unsigned char *)keys[tst].pub,
```
Comment (t8m) [inline test/evp_extra_test.c:4681]: Please do not use comma operator unnecessarily.

File: test/evp_extra_test.c [hunk 12]
```diff
@@ -4647,6 +4652,74 @@ const OPTIONS *test_get_options(void)
...
+        goto check_err;
+
+    if (EVP_DigestSign(ctx, NULL, &maclen, msg, sizeof(msg)) != 1)
+        goto check_err;
+
+    if (!TEST_ptr(mac = OPENSSL_malloc(maclen)))
+        goto err;
+
+    if (EVP_DigestSign(ctx, mac, &maclen, msg, sizeof(msg)) != 1)
+        goto check_err;
```
Comment (t8m) [inline test/evp_extra_test.c:4702]: Here I'd use:
```suggestion
    if (!TEST_int_eq(EVP_DigestSign(ctx, mac, &maclen, msg, sizeof(msg)), 0))
        goto err;
```

**Review by t8m — CHANGES_REQUESTED** [review]

File: test/evp_extra_test.c (hunk 8 above)
Comment (fffonion) [reply test/evp_extra_test.c:4627]: agreed, that's even better!

File: test/evp_extra_test.c (hunk 11 above)
Comment (fffonion) [reply test/evp_extra_test.c:4681]: i realized this was a typo after getting confused by this comment, good eyes!

**Review by t8m — APPROVED** [review]

File: test/evp_extra_test.c [hunk 13]
```diff
@@ -4647,6 +4650,73 @@ const OPTIONS *test_get_options(void)
...
+    const unsigned char msg[] = { 0x00, 0x01, 0x02, 0x03 };
+    int testresult = 0;
+    EVP_MD_CTX *ctx = NULL;
+    unsigned char *mac = NULL;
+    size_t maclen = 0;
+    unsigned char *pubkey;
+    size_t pubkeylen;
+
+    switch (keys[tst].type) {
+        case NID_X25519:
```
Comment (paulidale) [inline test/evp_extra_test.c:4668]: Style nit, we indent switch statements like this:
```c
    switch (keys[tst].type) {
    case NID_X25519:
    case NID_X448:
        return TEST_skip("signing not supported for X25519/X448");
    }
```

**Review by paulidale — APPROVED** [review]
Looks good.  Just one indentation style nit to fix and with that, it's approved.

**Comment by paulidale** [comment]
@t8m, you probably need to reapprove.

File: test/evp_extra_test.c (hunk 13 above)
Comment (fffonion) [reply test/evp_extra_test.c]: fixed!

**Review by paulidale — APPROVED** [review]

**Comment by fffonion** [comment]
Thanks everyone for reviewing!
As this issue also exists in 1.1.1 branch, should I open another PR targeting that branch?

**Comment by paulidale** [comment]
1.1.1 is strictly receiving security fixes only at this stage, so probably not.
Unless folks on the @openssl/otc think it should be.

**Comment by t8m** [comment]
I agree, this fix is not necessary for 1.1.1 given this is not really security fix.

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
Squashed and reworded commit message. Merged to master, 3.1, and 3.0 branches. Thank you for your contribution.

**Pull request closed without merging**